src/psyclone/domain/common/algorithm/psyir.py

```diff
@@ -128,54 +146,86 @@ def _validate_child(position, child):
+            if routine_root_name[0] == '"' and routine_root_name[-1] == '"' \
+               or \
+               routine_root_name[0] == "'" and routine_root_name[-1] == "'":
+                # fparser2 (issue #295) currently includes quotes as
+                # part of a string, so strip them out.
+                routine_root_name = routine_root_name[1:-1].strip()
+            routine_root_name = routine_root_name.replace(" ", "_")
+            # Check that the name is a valid routine name
+            pattern = re.compile(r"^[a-zA-Z]\w*$", re.ASCII)
+            if not pattern.match(routine_root_name):
+                raise TypeError(
+                    f"AlgorithmInvokeCall:_def_routine_root_name() the "
+                    f"(optional) name of an invoke must be a string "
+                    f"containing a valid name (with any spaces replaced by "
+                    f"underscores) but found '{routine_root_name}'.")
+        else:
+            routine_root_name = f"invoke_{self._index}"
+            if len(self.children) == 1:
+                # Add the name of the kernel if there is only one call
+                routine_root_name += "_" + self.children[0].name
         return routine_root_name
 
-    def create_psylayer_symbols(self):
-        '''If the PSy-layer routine and container symbols have not been
-        created, then create them. The names are based on the position
-        of this node (compared to other nodes of the same type) in the
-        PSyIR tree.
+    def create_psylayer_symbol_root_names(self):
+        '''If the PSy-layer routine and container root names have not been
+        created, then create them. The invoke root name is based on
+        the position of this node (compared to other nodes of the same
+        type) in the PSyIR tree. Note, we do not create and store
+        symbols, as the container name needs to be consistent between
+        different invoke calls and we have no way of knowing whether
+        one has already been created without the symbol being stored
+        in the symbol table, and we don't want to add anything related
+        to the lower level PSyIR to the symbol table before lowering.
 
         '''
-        if self.psylayer_routine_symbol:
-            # The language-level symbols have already been created
-            return
-
-        routine_root_name = self._def_routine_root_name()
-
-        symbol_table = self.scope.symbol_table
-        routine_name = symbol_table.next_available_name(
-            root_name=routine_root_name)
-
-        container_root_name = "{0}_mod".format(routine_name)
-        container_name = symbol_table.next_available_name(
-            root_name=container_root_name)
-
-        interface = ImportInterface(ContainerSymbol(container_name))
-        self.psylayer_routine_symbol = RoutineSymbol(
-            routine_name, interface=interface)
+        self.psylayer_routine_root_name = self._def_routine_root_name()
```

Comment:
I think it would be a good idea to only allow the name creation the first time:
```
if self.psylayer_routine_root_name is None:
    # Current implementation
```
Since there may already exist a matching PSy-layer routine if the create has been called before, and this could leave it orphan.

Reply:
I've protected the assignments and added tests. Note, I've done separate first time tests for both assignments as it feels more robust, even though it is not necessary.